combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/NodeBleDevice.kt

```diff
@@ -0,0 +1,117 @@
+package inc.combustion.framework.ble.device
+
+import android.bluetooth.BluetoothAdapter
+import android.util.Log
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.lifecycleScope
+import inc.combustion.framework.LOG_TAG
+import inc.combustion.framework.ble.scanning.CombustionAdvertisingData
+import inc.combustion.framework.ble.uart.meatnet.*
+import inc.combustion.framework.ble.uart.meatnet.NodeReadFeatureFlagsRequest
+import inc.combustion.framework.service.DebugSettings
+import inc.combustion.framework.service.DeviceConnectionState
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+internal class NodeBleDevice(
+    mac: String,
+    owner: LifecycleOwner,
+    private var nodeAdvertisingData: CombustionAdvertisingData,
+    adapter: BluetoothAdapter,
+    private var uart: UartBleDevice = UartBleDevice(mac, nodeAdvertisingData, owner, adapter)
+) {
+
+    protected val genericRequestHandler = UartBleDevice.MessageCompletionHandler()
+    protected val readFeatureFlagsRequest = UartBleDevice.MessageCompletionHandler()
+
+    companion object {
+        const val NODE_MESSAGE_RESPONSE_TIMEOUT_MS = 5000L
+    }
+
+    init {
+        processUartResponses()
+    }
+
+    val rssi: Int get() { return uart.rssi }
+    val connectionState: DeviceConnectionState get() { return uart.connectionState }
+    val isConnected: Boolean get() { return uart.isConnected.get() }
+    val isDisconnected: Boolean get() { return uart.isDisconnected.get() }
+    val isInRange: Boolean get() { return uart.isInRange.get() }
+    val isConnectable: Boolean get() { return uart.isConnectable.get() }
+
+    fun sendNodeRequest(request: GenericNodeRequest, callback: ((Boolean, Any?) -> Unit)?) {
+        val nodeRequest = request.toNodeRequest()
+        genericRequestHandler.wait(uart.owner, NODE_MESSAGE_RESPONSE_TIMEOUT_MS, null, callback)
+        sendUartRequest(nodeRequest)
+    }
+
+    fun sendFeatureFlagRequest(reqId: UInt?, callback: ((Boolean, Any?) -> Unit)?) {
+        val serialNumber = uart.serialNumber?: return
+        readFeatureFlagsRequest.wait(uart.owner, NODE_MESSAGE_RESPONSE_TIMEOUT_MS, reqId, callback)
+        sendUartRequest(NodeReadFeatureFlagsRequest(serialNumber, reqId))
+    }
+
+    var isInDfuMode: Boolean
+        get() = uart.isInDfuMode
+        set(value) {
+            uart.isInDfuMode = value
+        }
+
```

Comment:
I'm not sure how you're reading the serial number right now.  But I think you'll want to pull over [this code into this class](https://github.com/combustion-inc/combustion-android-ble/blob/de961f26b331eab8255c1b9298f46b285ea4516d/combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/ProbeBleDevice.kt#L100C1-L103C102) to be able to read the serial number of the node.

Reply:
That same code is below but with a different name. I'll update it so they are in sync.

Reply:
Done.